File: 2024/Day1/main.py
from collections import Counter
import sys

left: list[int] = []
right: list[int] = []

for line in sys.stdin:
    line = line.strip()
    if not line:
        continue

    parts = line.split()
    if len(parts) == 2:
        left.append(int(parts[0]))
        right.append(int(parts[1]))

left.sort()
right.sort()

total_distance = 0
for i in range(len(left)):
    total_distance += abs(left[i] - right[i])

print(f'Part 1: {total_distance}')

right_counts = Counter(right)
similatiry_score = 0
for num_in_left in left:
    similatiry_score += num_in_left * right_counts[num_in_left]

print(f'Part 2: {similatiry_score}')

File: 2024/Day1/main.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <unordered_map>
#include <vector>

int main() {
    std::ifstream file("../input.txt");
    std::vector<int> left, right;
    int a, b;

    while (file >> a >> b) {
        left.push_back(a);
        right.push_back(b);
    }

    std::sort(left.begin(), left.end());
    std::sort(right.begin(), right.end());

    long long total_distance = 0;
    for (size_t i = 0; i < left.size(); i++) {
        total_distance += std::abs(left[i] - right[i]);
    }

    std::cout << "Part 1: " << total_distance << std::endl;

    std::unordered_map<int, int> right_counts;
    for (int num : right) {
        right_counts[num]++;
    }

    long long similiatiry_score = 0;
    for (int num : left) {
        similiatiry_score += num * right_counts[num];
    }

    std::cout << "Part 2: " << similiatiry_score << std::endl;

    return 0;
}
